File: src/types.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { createDiagnostic, Diagnostics, formatMessage } from "./diagnostics.ts";
import { SyntaxKind } from "./types.ts";
import {
  isKeyword,
  isModifierKeyword,
  isPrimitiveTypeKeyword,
  textToKeyword,
  tokenToString,
} from "./utilities.ts";

test("range markers are ordered", () => {
  expect(SyntaxKind.FirstLiteralToken <= SyntaxKind.LastLiteralToken).toBe(true);
  expect(SyntaxKind.FirstPunctuation <= SyntaxKind.LastPunctuation).toBe(true);
  expect(SyntaxKind.FirstAssignment <= SyntaxKind.LastAssignment).toBe(true);
  expect(SyntaxKind.FirstKeyword <= SyntaxKind.LastKeyword).toBe(true);
  expect(SyntaxKind.FirstReservedWord <= SyntaxKind.LastReservedWord).toBe(true);
  expect(SyntaxKind.FirstTypeNode <= SyntaxKind.LastTypeNode).toBe(true);
  expect(SyntaxKind.FirstStatement <= SyntaxKind.LastStatement).toBe(true);
  expect(SyntaxKind.FirstExpression <= SyntaxKind.LastExpression).toBe(true);
});

test("every keyword kind falls inside the keyword range", () => {
  for (const kind of textToKeyword.values()) {
    expect(kind >= SyntaxKind.FirstKeyword && kind <= SyntaxKind.LastKeyword).toBe(true);
    expect(isKeyword(kind)).toBe(true);
  }
});

test("keyword text round-trips through textToKeyword and tokenToString", () => {
  for (const [text, kind] of textToKeyword) {
    expect(textToKeyword.get(text)).toBe(kind);
    expect(tokenToString(kind)).toBe(text);
  }
});

test("Identifier is not a keyword", () => {
  expect(isKeyword(SyntaxKind.Identifier)).toBe(false);
});

test("tokenToString returns spellings for punctuation", () => {
  expect(tokenToString(SyntaxKind.OpenBraceToken)).toBe("{");
  expect(tokenToString(SyntaxKind.GreaterThanGreaterThanGreaterThanToken)).toBe(">>>");
  expect(tokenToString(SyntaxKind.ArrowToken)).toBe("->");
  expect(tokenToString(SyntaxKind.Identifier)).toBeUndefined();
});

test("modifier and primitive predicates", () => {
  expect(isModifierKeyword(SyntaxKind.PublicKeyword)).toBe(true);
  expect(isModifierKeyword(SyntaxKind.ClassKeyword)).toBe(false);
  expect(isPrimitiveTypeKeyword(SyntaxKind.IntKeyword)).toBe(true);
  expect(isPrimitiveTypeKeyword(SyntaxKind.VoidKeyword)).toBe(false);
});

test("diagnostics format placeholders and locations", () => {
  expect(formatMessage(Diagnostics._0_expected, [";"])).toBe("';' expected.");
  const d = createDiagnostic(5, 3, Diagnostics._0_expected, "}");
  expect(d.pos).toBe(5);
  expect(d.end).toBe(8);
  expect(d.messageText).toBe("'}' expected.");
  expect(d.code).toBe(Diagnostics._0_expected.code);
});

File: togo/internal/compiler/types_test.go
package compiler

import "testing"

// Port of src/compiler/types.test.ts (the range-marker, keyword round-trip,
// punctuation, predicate and diagnostic-format cases; the SyntaxKind name-table
// completeness and keyword-range cases live in kinds_test.go).

func TestRangeMarkersOrdered(t *testing.T) {
	pairs := [][2]SyntaxKind{
		{FirstLiteralToken, LastLiteralToken},
		{FirstPunctuation, LastPunctuation},
		{FirstAssignment, LastAssignment},
		{FirstKeyword, LastKeyword},
		{FirstReservedWord, LastReservedWord},
		{FirstTypeNode, LastTypeNode},
		{FirstStatement, LastStatement},
		{FirstExpression, LastExpression},
	}
	for _, p := range pairs {
		if p[0] > p[1] {
			t.Errorf("range marker out of order: %v > %v", p[0], p[1])
		}
	}
}

func TestEveryKeywordInRange(t *testing.T) {
	for _, kind := range textToKeyword {
		if kind < FirstKeyword || kind > LastKeyword || !IsKeyword(kind) {
			t.Errorf("keyword kind %v not inside the keyword range", kind)
		}
	}
}

func TestKeywordTextRoundTrips(t *testing.T) {
	for text, kind := range textToKeyword {
		if got := tokenToString(kind); got != text {
			t.Errorf("tokenToString(%v) = %q, want %q", kind, got, text)
		}
	}
}

func TestIdentifierNotKeyword(t *testing.T) {
	if IsKeyword(Identifier) {
		t.Error("Identifier should not be a keyword")
	}
}

func TestTokenToStringPunctuation(t *testing.T) {
	if got := tokenToString(OpenBraceToken); got != "{" {
		t.Errorf("OpenBraceToken = %q, want {", got)
	}
	if got := tokenToString(GreaterThanGreaterThanGreaterThanToken); got != ">>>" {
		t.Errorf("got %q, want >>>", got)
	}
	if got := tokenToString(ArrowToken); got != "->" {
		t.Errorf("got %q, want ->", got)
	}
	if got := tokenToString(Identifier); got != "" {
		t.Errorf("Identifier spelling = %q, want empty", got)
	}
}

func TestModifierAndPrimitivePredicates(t *testing.T) {
	if !isModifierKeyword(PublicKeyword) || isModifierKeyword(ClassKeyword) {
		t.Error("modifier predicate wrong")
	}
	if !isPrimitiveTypeKeyword(IntKeyword) || isPrimitiveTypeKeyword(VoidKeyword) {
		t.Error("primitive predicate wrong")
	}
}

func TestDiagnosticsFormat(t *testing.T) {
	if got := FormatMessage(Diagnostics.Expected0, ";"); got != "';' expected." {
		t.Errorf("formatMessage = %q", got)
	}
	d := CreateDiagnostic(5, 3, Diagnostics.Expected0, "}")
	if d.Pos != 5 || d.End != 8 || d.MessageText != "'}' expected." || d.Code != Diagnostics.Expected0.Code {
		t.Errorf("diagnostic = %+v", d)
	}
}
